### PR TITLE
Docstrings, de-duplicate EAMixin/DatetimeLikeIndex __new__ code

### PR DESCRIPTION
**Pull request by jbrockmendel**
There's a lot of duplication in the constructors and constructor-helpers.  This starts to whittle that down, writes some docstrings long the way.

Also use `ExtensionOpsMixin` to define comparison operators on the EAMixin classes.

We determined that the DatetimeIndex._local_timestamps method had an unecessary monotonicy check, so took that out.

### CODE REVIEW

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -634,6 +634,11 @@ class ExtensionOpsMixin(object):
     """
     A base class for linking the operators to their dunder names
     """
+
+    @classmethod
```
Comment (jreback) [inline pandas/core/arrays/base.py]: ideally should add doc-strings & to the top-level docs, as this is API

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -551,3 +557,97 @@ def validate_periods(periods):
...
+
+    Raises
+    ------
+    ValueError : on tzinfo mismatch
+    """
+    if dtype is not None:
+        try:
+            dtype = DatetimeTZDtype.construct_from_string(dtype)
+            dtz = getattr(dtype, 'tz', None)
+            if dtz is not None:
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can you do this ?

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine. if you can just update the docs in arrays/base.py when you add api there

File: pandas/core/arrays/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/base.py]: OK.  I think I'll revert this part of the PR as out-of-scope and do the docs bit separately.

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -23,10 +24,13 @@
     is_timedelta64_dtype,
     is_object_dtype)
 from pandas.core.dtypes.generic import ABCSeries, ABCDataFrame, ABCIndexClass
+from pandas.core.dtypes.dtypes import DatetimeTZDtype
 
 import pandas.core.common as com
 from pandas.core.algorithms import checked_add_with_arr
 
+from .base import ExtensionOpsMixin
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can you do absolute imports

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -464,7 +468,10 @@ def _addsub_offset_array(self, other, op):
                       "{cls} not vectorized"
                       .format(cls=type(self).__name__), PerformanceWarning)
 
-        res_values = op(self.astype('O').values, np.array(other))
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: is it worth a branch here in is_extension_dtype?

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Tried it, causes import-time errors

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: I think `values_from_object` is a pretty clean solution to this.

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: i see, we import in __init__ all of the extension arrays which import this. hmm.

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: can you update this

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Already done

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: hmm, does tz_compare not take None as an arg? hmm make should add the check there? as I don't think we check for None's elsewhere (not really sure about this)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: None-tz comparisons are a pretty special case that get handled separately.  I’m OK with tzcompare not taking None.

**Comment by jbrockmendel** [comment]
@jreback since we're on a roll I'd like to push this through, get started on the next steps over the weekend.  Thoughts?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
sure. my comments were minor.

**Pull request merged**